### ISSUE TITLE
[Discovery] Relationship between the VA Forms DB/Lighthouse API/Drupal

### ISSUE THREAD

**Issue by MickinSahni** [department-of-veterans-affairs/va.gov-team#5702]
## Background:
We rely on the VA Forms API to power our new Find VA Forms feature.
Find VA Forms returns individual VA form pages with form-specific data/content for users. 
The source of truth for forms is the VA Forms DB. 
However, the DB does not contain every field we need to surface to users. 

Let's answer these questions: 
- What additional data do we need to store?
- Is this preventing us from launching the new form pages?
- Where should this data be maintained?
- How should this new data be maintained and what's Drupal's role?

## Acceptance Criteria: 
* Decision made on where the form data is going to be maintained and how it will be manipulated e.g. via Drupal
* Decision made on who will make requisite changes to the DB, API, and/or Drupal
* Decision on how the front-end will access this form data
* New tickets created under epic #5590 for subsequent implementation
* The [comment](https://github.com/department-of-veterans-affairs/va.gov-team/issues/5277#issuecomment-580782604) here is addressed in any decision we make re: keywords 

Notes: there is a ticket with the facilities team to actually create the Drupal template for the form landing page

**Comment by ncksllvn** [comment]
This doesn't directly answer the questions in the issue, but here's a summary from the call on February 12th, attended by Public Websites, CMS, and Lighthouse team members -

### Forms Search
- The Forms Search page should have parity with the legacy VA Forms search.
    - This may mean we need access to the VA Forms database before launch, in order to use the `keywords` table.
- We have the data fields we need for launching an MVP of the Forms Search. This means there aren't complicated relationships for this product, although the architecture of the Form Detail pages will likely carry impact.

### Form Detail pages
[Diagrams of various forms of relationships across systems](https://whimsical.com/Bn5zRk2PDzuSjpu1d5q4Pu)

The VA Forms Database, essentially exposed through the Lighthouse API, doesn't contain every field we need to render the Form Detail pages. The additional field - associated forms, descriptions, relevant links, benefits hubs, etc - should be housed in Drupal, so that we can easily maintain and extend the data. However, any fields housed in the VA Form DB should originate from there, so that it remains the source of truth for VA Forms.
 
We could consume the Lighthouse API into Drupal, add the Drupal-housed fields, then pull all of that data in the deployment of vets-website to generate static HTML pages. This aligns with how facility pages are created -

![image](https://user-images.githubusercontent.com/1915775/74384139-a6748300-4dbe-11ea-9858-f11c3c02cfd3.png)

__However, these Drupal-housed data fields would also be valuable to the Lighthouse API itself, in order to benefit other consumers of the Lighthouse API.__ This approach would not benefit the Lighthouse API, which would be a major benefit for the Lighthouse team's API.

What we're considering then is approach where Drupal pulls from the VA Forms database. Drupal then adds its additional data fields, exposing the complete data through its GraphQL API. The Lighthouse API can then pull from Drupal using that GraphQL API. 

![image](https://user-images.githubusercontent.com/1915775/74384886-3f57ce00-4dc0-11ea-9669-1616bd5e6946.png)


Since the data flow is one-way and benefits the Lighthouse API as well, this is the most ideal solution, but would not be completely actionable until we have access to the VA Forms DB. If we don't have access to the VA Forms DB in the short term, we may consider another path, that includes relationships to and from the Lighthouse API and Drupal. This would introduce added tech debt though, since the best solution is the one described above.

Eventually, it would also be ideal if the VA Forms DB were to be deprecated, so that the source of truth for forms can be completely internal to Drupal -

![image](https://user-images.githubusercontent.com/1915775/74384784-0e779900-4dc0-11ea-8785-df018272a3e6.png)

**Comment by ncksllvn** [comment]
Updated diagram to reflect technical architecture and team ownership of tasks - 

![image.png](https://images.zenhubusercontent.com/59cb9de1b0222d5de47953d8/86aa0d2d-194a-4194-804f-e308df13b65f)

**Comment by ncksllvn** [comment]
This does raise the question for me...Is it true that Public Websites will see a Drupal engineer join? Because they could be really useful lol.

**Comment by MickinSahni** [comment]
> This does raise the question for me...Is it true that Public Websites will see a Drupal engineer join? Because they could be really useful lol.

cc @jenniferlee-dsva fya re: other potential Drupal needs

**Comment by ncksllvn** [comment]
> What additional data do we need to store?

These fields exist already in the __VA Forms Database__, so we should pull them from there once we have access -

1. __FormNum__
1. __FormTitle__
1. __FileName__
1.  __FormType__
    - This is required for constructing the URL with the FileName field above.
1. __OnlineFormURL__
    - Uncertain about the accuracy of this value. We may not be interested.
1. __revision_date__
1. __IntranetOnly__ 
    - Must be `false` for our purposes, since we're only interested in public-facing forms)
1. __Keywords__
   - Defined via the `form_keywords` table (maps between form IDs to keyword IDs) and the `keyword` table

In __Drupal__, we should store these additional fields. This was largely broken down by the latest [mockups](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/find-va-forms-redesign/VA-form-master-detail-LP-patterns-mockups-v7.pdf) and this [VA Forms MVP Redesign New DB Fields spreadsheet](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/find-va-forms-redesign/VA-forms-mvp-redesign-new-db-field-010620.xlsm).

1. __Related To__
    - Possible values: Burials and memorials benefits, Careers and employment benefits, Disability benefits, Education and training benefits, Health care benefits, Housing assistance benefits, Insurance benefits, Pension benefits, Records, Employee forms (INTERNAL), VA medical centers, Family member benefits, Service member benefits
    - Can have more than one value
    - We have been talking about storing this as an addition to the VA Forms Database, but it seems like any new fields should be housed in Drupal.
1. __Online Tool__
    - Values: True/false (or yes/no)
1.  __URL to online tool__
    - Exists only if Online Tool is true. Perhaps we could just leave this field empty to indicate there isn't an Online Tool, and not require the Online Tool field.
1. __When to use this form__
   - Text description
1. __Related forms and instructions__
    - A list of forms. Each form has - heading, description, and URL
1. __Helpful links__
    - A list of links. Each link has - heading, description, and URL
    - Should have a default list of links, but can also be customized.

> Is this preventing us from launching the new form pages?

Largely, yes, but we could propose an interim solution. If we have a spreadsheet that contains all of the new fields and values that we want to see eventually stored into Drupal, we could temporarily leverage that as static data directly from the front-end. Then, when a user navigates to a form detail page, we could load that form's data from the Lighthouse API to get the form number, form name, download URL, etc., and then pull the rest of the form's data out of that static data in the front-end to imitate the complete data structure we should receive from Drupal. This would allow us to start building the front-end. Then, once the work with defining those fields in Drupal is finished, we could update the front-end to remove the Lighthouse API call and static data to instead leverage Drupal.

> Where should this data be maintained?

Existing fields in the VA Forms Database should continue to live there, but all additional fields should live and be maintained via Drupal.

> How should this new data be maintained and what's Drupal's role?

Fields from the VA Forms Database should be "piped" into Drupal. When editing a form detail page using the form number, these values should be presented as uneditable input fields. Fields that are local to Drupal - Related To, Online Tool, etc - should exist as editable input fields in a typical Drupal editorial experience.

Then, the Lighthouse API can pull from Drupal to essentially cache the values for the form data. The front-end can then query the Lighthouse API from the browser. The advantages are -

1. Drupal is defined as the source of truth for VA forms data, so Public Websites has content-ownership
1. The Lighthouse API is enhanced by the additional fields made available from Drupal
1. The front-end can be built as a typical React app that simply pulls data from an API using the form number

**Comment by ncksllvn** [comment]
> Decision made on where the form data is going to be maintained and how it will be manipulated e.g. via Drupal

✔️ Drupal!

> Decision made on who will make requisite changes to the DB, API, and/or Drupal

I think that's reflected in the [flow chart](https://github.com/department-of-veterans-affairs/va.gov-team/issues/5702#issuecomment-587579257) above. It appears that most of the work will fall into Public Website's court as far as Drupal and the front-end, but I'm not sure if PW should be leading the integration of the Drupal forms data into Lighthouse. That seems like Lighthouse territory.

> Decision on how the front-end will access this form data

✔️ As a typical React app pulling from an API (Lighthouse in our case) although we have a potential interim solution available by using a static data structure to substitute for Drupal until Drupal implementation is complete.

> New tickets created under epic #5590 for subsequent implementation

❌ Todo! I'll scope out an idea for ticket breakdown, but we probably need to huddle before creating the tickets.

> The comment [here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/5277#issuecomment-580782604) is addressed in any decision we make re: keywords

✔️ This is related to the forms search (not the form detail pages, as this ticket refers) but from the call on February 7th, it sounds like that we want complete parity with the legacy forms search before we can roll out our modernized forms search. So, we'll likely be blocked on launching the forms search until the Lighthouse API for searching forms has access to the VA Forms DB instead of scraping the legacy webpage, because the Lighthouse API will need the ability to integrate  `keywords` into its search algorithm.

**Comment by jenniferlee-dsva** [comment]
Re above comments... 
Comment re [Drupal engineer](https://github.com/department-of-veterans-affairs/va.gov-team/issues/5702#issuecomment-587579735) - yes, we are. It may be long term via the CMS team and scrubbing in to a product team; short term, near future, Dave Conlon's VAMC team may 'loan' us their Drupal engineer . See ticket #5653. 

Comment above re search parity and keywords: this is on my item to discuss on tomorrow's call with @ncksllvn.  

Added 2 comments in [message re form db fields](https://github.com/department-of-veterans-affairs/va.gov-team/issues/5702#issuecomment-587791167)

**Comment by ncksllvn** [comment]
1. Can we launch the forms search with using term-based search via form titles? No need to access the keywords field in the database? Our team can implement the DB query, but without access to the VA Forms DB, we can't leverage keywords to create true parity w/the legacy forms experience. 

   - We may be able to launch with only term-based search using the form name, but we should evaluate after implementation of #5693.

1. Where should we store RelatedTo and OnlineTool/URL? Who do we expect to maintain the RelatedTo and OnlineTool/URL? Us, or the form managers?
    
    -  Can we confirm when forms are added/modified in the forms DB during the Drupal migration?
         - Check w/CMS team - if new/edited forms are flagged in Drupal, then it could be useful to maintain this data in Drupal.

2. Do we have any realistic ability to move the forms management (database) entirely into Drupal at some point?
    - Potentially in Q2/3. There is varied interest, with concern mostly around what happens to intranet-only forms, and if form managers will find themselves needing to update forms in two locations - public-facing (Drupal) vs intranet (VA Forms DB.)

**Comment by MickinSahni** [comment]
2020-02-19  PW Standup
- nick to reach out about additional questions
- will finalize what's left for this discovery in pt. 2 grooming

**Comment by MickinSahni** [comment]
2020-02-20 PW Standup
- Last piece of this ticket is to determine where RelatedTo and OnlineTool/URL will live—VA Forms DB or Drupal DB
  - first indication was VA Forms DB so all new forms can have this filled, but would not give us ability to edit this field unless it's in the Drupal DB
- outcome of this ticket will be a bulleted list of the tickets necessary—task + team/owner

**Comment by MickinSahni** [comment]
- we've decided our path forward with respect to Drupal changes and Lighthouse updates
- concern: we have **not** decided how those two systems will be integrated i.e. seems like Lighthouse territory to plug into Drupal. 
  - New ticket: 'Determine Lighthouse-Drupal integration approach' to the form detail page epic - cc @AkucherGCIO

**Comment by ncksllvn** [comment]
On call with @AkucherGCIO, thinking how to break this into tickets - 

1. Get credentials for the VA Forms Database - is there ATO involved?
1. [Drupal] Create a nightly migration from VA Forms DB into Drupal
1. [Drupal] Create a "node -type" for Form Detail pages
      - Custom Form Landing Pages = Top 30 forms, which contains extra fields versus the less-popular forms
      - Can we also add fields "RelatedTo" and "Online Tool" for _all_ forms? Can we be notified when a new form is added into the VA Forms DB?
1. [Lighthouse] Being consuming form data from the Lighthouse Forms API
    - Use Drupal's GraphQL API to pull form data from Drupal into the Lighthouse Forms local database
1. [Front-End] Begin consuming Form-Detail data from Lighthouse API

**Comment by AkucherGCIO** [comment]
2020-26-02
@ncksllvn  - I will create the tickets above and get them ready for grooming.  Do we want to close this out as part of Sprint 15 ?

**Comment by ncksllvn** [comment]
@AkucherGCIO sounds good! I think we can close it now. The main things I'd emphasize as takeaways -

1. We are essentially blocked from starting the Drupal work required here until we have access to the VA Forms DB.
1. We are still unsure of whether new database fields - the `Related To` and `URL to Online Tool` should originate in the VA Forms DB as new fields, or be created in Drupal. 
1. We _can_ consider an interim front-end solution to help expedite front-end development and get to a "showable" state for the Form Detail pages.

**Comment by kevwalsh** [comment]
Example experience (still could use some UX love) to emulate.
![Edit_VAMC_facility_Lebanon_VA_Medical_Center___VA_gov_CMS](https://user-images.githubusercontent.com/643678/75475649-237b2d00-5967-11ea-96f0-38c57e774aa6.jpg)

**Comment by ncksllvn** [comment]
CC: @beeyayjay

**Comment by beeyayjay** [comment]
@kevwalsh, could I get a link to that " good feedback from VHA facility editors" you mentioned?
